apollo-ios/Design/swift-identifiable.md

```diff
@@ -77,18 +95,32 @@ The protocol requires that the identity is accessible through a public field nam
 public var id: String { self.uuid }
 ```
 
-## Naming conflicts
+### Naming conflicts
 
 If the identity field is not called `id`, but another field called `id` is present in the selection, a custom getter cannot be added. Swift does not support using another field to handle the conformance.
 
 In this case, a conformance to Identifiable SHOULD NOT be generated. Codegen should emit a warning without stopping the generation process.
 
+## Caching behavior
+
+A scope of `SELECTION` does not allow the identifier to be used as a caching key. Clients MAY use other mechanisms to determine if and how to cache the object.
+
+An identifier with scope `TYPE` can be combined with the `__typename` field to generate a caching key that's unique for the GraphQL Service.
+
+Identifiers with a scope of `SERVICE` or `GLOBAL` can be directly used as a caching key.
+
 # Alternatives
 
 ## Automatic conformance to Identifiable
 
 The code generator could be updated to always emit a conformance to Identifiable if a scalar `id` field is present in the selection set, removing the need for a custom directive. This was suggested in Pull Request [#548](https://github.com/apollographql/apollo-ios-dev/pull/548).
 
-## Declaring identity scope
+## Apollo Kotlin's @typePolicy directive
+
+Apollo Kotlin has [custom directives](https://www.apollographql.com/docs/kotlin/caching/declarative-ids) that allows for client authors to specify the caching key though pure GraphQL:
+
+```graphql
+extend type Book @typePolicy(keyFields: "id")
+```
 
-The directive could be expanded to declare the scope of an identity (e.g. unique across the entire API or only one specific Database table). This makes the directive beneficial outside of Swift code generation, but it doesn't change the behavior of the Identifiable protocol. The [Swift documentation](https://developer.apple.com/documentation/swift/identifiable) states that the scope and duration of an identity is unspecified, so modifying the scope would not lead to a difference in the generated code.
+This could be used for Identifiable conformance if a single field is provided, and caching behavior could also be matched with Apollo Kotlin. A disadvantage is that this directive is only used inside type extensions, and can't be used to mark fields directly in a query.
```

Comment:
If `@typePolicy` was used could the Identifiable `id` value be a hash value of all specified key fields? I don't have an answer for not being able to use `@typePolicy` in operations though.

Reply:
If this were possible then maybe `Identifiable` conformance could just be a codegen configuration option (`generateIdentifiableConformance: Bool`) and the hashed type policy key values are used for `id` value?

Not great that there would be two places for configuration but it does highlight that `Identifiable` is a side-effect of object identity. Maybe that's not a bad requirement, plus it wouldn't need to be used in operations then?

Reply:
> could the Identifiable id value be a hash value of all specified key fields?

You're right, I forgot that a tuple of `Hashable` objects is also `Hashable` itself, which makes it valid for use with `Identifiable`.

Reply:
> If this were possible then maybe Identifiable conformance could just be a codegen configuration option (generateIdentifiableConformance: Bool)

When the identity is established as a single field named `id`, there is no drawback to always adding Identifiable conformance.

If a custom getter for `id` has to be added for conformance, then following a configuration option seems like a good idea. This would also make it less surprising when Codegen suddenly emits a warning about a getter being unable to be generated because of naming conflicts.

Reply:
> When the identity is established as a single field named id, there is no drawback to always adding Identifiable conformance.

Even if it is a single field named `id` I think we'd still prefer configuration over convention. That configuration option can have a default of `true` if not supplied explicitly = less auto-magic behaviour.